Xamarin.PropertyEditing.Mac/Controls/CombinablePropertyEditor.cs

```diff
@@ -0,0 +1,142 @@
+using System;
+using System.Collections;
+using System.Diagnostics;
+using System.ComponentModel;
+
+using Foundation;
+using AppKit;
+using CoreGraphics;
+
+using Xamarin.PropertyEditing.ViewModels;
+using Xamarin.PropertyEditing.Mac.Resources;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Xamarin.PropertyEditing.Mac
+{
+	internal class CombinablePropertyEditor<T>
+		: PropertyEditorControl
+	{
+		public CombinablePropertyEditor ()
+		{
+			base.TranslatesAutoresizingMaskIntoConstraints = false;
+
+			UpdateTheme ();
+		}
+
+		public override NSView FirstKeyView => firstKeyView;
+		public override NSView LastKeyView => lastKeyView;
+
+		internal new CombinablePropertyViewModel<T> ViewModel
+		{
+			get { return (CombinablePropertyViewModel<T>)base.ViewModel; }
+			set { base.ViewModel = value; }
+		}
+
+		List<NSButton> combinableList = new List<NSButton> ();
+		NSView firstKeyView;
+		NSView lastKeyView;
+
+		protected override void HandleErrorsChanged (object sender, DataErrorsChangedEventArgs e)
+		{
+			UpdateErrorsDisplayed (ViewModel.GetErrors (e.PropertyName));
+		}
+
+		protected override void SetEnabled ()
+		{
+			foreach (var item in combinableList) {
+				item.Enabled = ViewModel.Property.CanWrite;
+			}
+		}
+
+		protected override void UpdateErrorsDisplayed (IEnumerable errors)
+		{
+			if (ViewModel.HasErrors) {
+				SetErrors (errors);
+			} else {
+				SetErrors (null);
+				SetEnabled ();
+			}
+		}
+
+		protected override void OnViewModelChanged (PropertyViewModel oldModel)
+		{
+			combinableList.Clear ();
+
+			const float controlHeight = 22;
+
+			// Set our new RowHeight
+			RowHeight = (ViewModel.Choices.Count * controlHeight);
+
+			float top = controlHeight;
+			foreach (var item in ViewModel.Choices) {
+				var BooleanEditor = new NSButton (new CGRect (4, RowHeight - top, Frame.Width - 4, controlHeight)) {
+					AllowsMixedState = true,
+					ControlSize = NSControlSize.Small,
+					Font = NSFont.FromFontName (DefaultFontName, DefaultFontSize),
+					Title = item.Name,
+					TranslatesAutoresizingMaskIntoConstraints = false,
+				};
+				BooleanEditor.SetButtonType (NSButtonType.Switch);
+				BooleanEditor.Activated += SelectionChanged;
+
+				AddSubview (BooleanEditor);
+				combinableList.Add (BooleanEditor);
+				top += controlHeight;
+			}
+
+			// Set our tabable order
+			firstKeyView = combinableList.First ();
+			lastKeyView = combinableList.Last ();
+
+			base.OnViewModelChanged (oldModel);
+		}
+
+		protected override void UpdateValue ()
+		{
+			foreach (var item in combinableList) {
+				if (ViewModel.Choices.Count > 0) {
+					var flag = ViewModel.Choices.FirstOrDefault (x => x.Name == item.Title);
+					if (flag.IsFlagged.HasValue) {
+						item.State = flag.IsFlagged.Value ? NSCellStateValue.On : NSCellStateValue.Off;
+						item.AllowsMixedState = false;
```

Comment:
You're disabling mixed state here, but not re-enabling it below. A value that differs when a second object selected will not show as mixed like it should.

Reply:
I thought that once the value as set (either ticked or unticked), it could never be null again, hence why it could never be mixed again?

Reply:
The initial state of the NSButton is always mixed, when the button is first created. So by the time UpdateValue is called if it HasValue is false it will already be in a Mixed State. I could set the State to Mixed on instantiating the NSButton and avoid this else.

Reply:
You're missing the point, the state can become mixed _after_ first setup:
> A value that differs when a second object selected will not show as mixed like it should.

> ... another case is if a second object is selected that an individual flag is set for one but not the other.

If you select the first mock, select flag 1, unselect it, select the second mock and select flag 2 and then select the first mock they should both show mixed, but they both show checked.